.github/workflows/ci.yml

```diff
@@ -15,30 +15,250 @@ jobs:
-      uses: actions/cache@v4
-      with:
-        path: |
-          ~/.gradle/caches
-          ~/.gradle/wrapper
-        key: ${{ runner.os }}-gradle-${{ hashFiles('**/*.gradle*', '**/gradle-wrapper.properties') }}
-        restore-keys: |
-          ${{ runner.os }}-gradle-
-          
-    - name: Make gradlew executable
-      run: chmod +x ./gradlew
-      
-    - name: Run tests
-      run: ./gradlew test
-      
-    - name: Run build
-      run: ./gradlew build
\ No newline at end of file
+      - name: Checkout code
+        uses: actions/checkout@v4
+
+      - name: Set up JDK ${{ matrix.java-version }}
+        uses: actions/setup-java@v4
+        with:
+          java-version: ${{ matrix.java-version }}
+          distribution: 'temurin'
+
+      - name: Cache Gradle dependencies
+        uses: actions/cache@v4
+        with:
+          path: |
+            ~/.gradle/caches
+            ~/.gradle/wrapper
+          key: ${{ runner.os }}-gradle-${{ hashFiles('**/*.gradle*', '**/gradle-wrapper.properties') }}
+          restore-keys: |
+            ${{ runner.os }}-gradle-
+
+      - name: Make gradlew executable
+        run: chmod +x ./gradlew
+
+      - name: Run tests
+        run: ./gradlew test
+
+      - name: Run build
+        run: ./gradlew build
+
+  benchmark:
+    runs-on: ubuntu-latest
+    needs: test
+    # Add explicit permissions for commenting on PRs
+    permissions:
+      contents: read
+      pull-requests: write
+      issues: write
+    # Only run benchmarks on main branch pushes and PRs to main to avoid excessive CI time
+    if: github.ref == 'refs/heads/main' || github.base_ref == 'main'
+
+    steps:
+      - name: Checkout code
+        uses: actions/checkout@v4
+
+      - name: Set up JDK 11
+        uses: actions/setup-java@v4
+        with:
+          java-version: '11'
+          distribution: 'temurin'
+
+      - name: Cache Gradle dependencies
+        uses: actions/cache@v4
+        with:
+          path: |
+            ~/.gradle/caches
+            ~/.gradle/wrapper
+          key: ${{ runner.os }}-gradle-${{ hashFiles('**/*.gradle*', '**/gradle-wrapper.properties') }}
+          restore-keys: |
+            ${{ runner.os }}-gradle-
+
+      - name: Make gradlew executable
+        run: chmod +x ./gradlew
+
+      - name: Create benchmark results directory
+        run: mkdir -p benchmark-results
+
+      - name: Run serialization benchmarks
+        run: |
+          ./gradlew jmhRunSerializationBenchmarks
+        continue-on-error: true
+
+      - name: Run deserialization benchmarks
+        run: |
+          ./gradlew jmhRunDeserializationBenchmarks
+        continue-on-error: true
+
+      - name: Run field access benchmarks
+        run: |
+          ./gradlew jmhRunFieldAccessBenchmarks
+        continue-on-error: true
+
+      - name: Run size comparison benchmarks
+        run: |
+          ./gradlew jmhRunSizeComparisonBenchmarks
+        continue-on-error: true
```

Comment:
I don't think we should run the comparison benchmarks on CI - they can be run on demand since we don't expect the performance of other systems to change when we commit to Imprint and the Imprint-specific benchmarks should catch regressions (which is the point of running things on each PR).

Not only does this slow down the PR builds, but I think there's some limit to the free GH plan on how many minutes you can run workflows for (though I need to double check that, it may only apply to private repos)